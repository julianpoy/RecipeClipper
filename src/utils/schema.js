import * as self from './schema';

const longestStringIn = (strArray) => strArray.reduce((max, match) => (match.length > max.length ? match : max), '');

export const getRecipeSchemasFromDocument = () => {
  const schemas = [...document.querySelectorAll('script[type="application/ld+json"]')]
    .map((schema) => {
      try {
        return JSON.parse(schema.innerText);
      } catch (e) {
        // Do nothing
      }

      return null;
    })
    .filter((e) => e);

  const recipeSchemas = schemas.map((schema) => {
    // Schemas that evaluate to a graph
    if (schema['@graph']) {
      return schema['@graph'].find((subSchema) => subSchema['@type'] === 'Recipe');
    }

    // Schemas that are directly embedded
    if (schema['@type'] === 'Recipe') return schema;

    // Schemas that evaluate to an array
    if (schema.length && schema[0]) {
      return schema.find((subSchema) => subSchema['@type'] === 'Recipe');
    }

    return null;
  }).filter((e) => e);

  return recipeSchemas;
};

let schemas;
export const getPropertyFromSchema = (propName) => {
  if (!schemas) schemas = self.getRecipeSchemasFromDocument();

  const foundSchema = schemas.find((schema) => schema[propName]) || {};

  return foundSchema[propName] || null;
};

export const getImageSrcFromSchema = () => {
  const images = self.getPropertyFromSchema('image');
  if (!images) return '';

  let imageSrc;
  if (typeof images === 'string') imageSrc = images;
  else if (typeof images[0] === 'string') [imageSrc] = images;

  if (imageSrc) {
    try {
      const url = new URL(imageSrc);

      if (url.protocol === 'http:' || url.protocol === 'https:') return imageSrc;
    } catch (_) {
      // Do nothing
    }
  }

  return '';
};

export const getTitleFromSchema = () => {
  const title = self.getPropertyFromSchema('name');

  if (typeof title === 'string') return title;

  return '';
};

export const getDescriptionFromSchema = () => {
  const description = self.getPropertyFromSchema('description');

  if (typeof description === 'string') return description;

  return '';
};

export const getYieldFromSchema = () => {
  const recipeYield = self.getPropertyFromSchema('recipeYield');
  if (!recipeYield) return '';

  if (typeof recipeYield === 'string') return recipeYield;
  if (typeof recipeYield[0] === 'string') return longestStringIn(recipeYield);

  return '';
};

export const getInstructionsFromSchema = () => {
  const instructions = self.getPropertyFromSchema('recipeInstructions');
  if (!instructions) return '';

  if (typeof instructions === 'string') return instructions;
  if (typeof instructions[0] === 'string') return instructions.join('\n');
  if (instructions[0] && typeof instructions[0].text === 'string') {
    return instructions.map((instruction) => instruction.text).join('\n');
  }

  return '';
};

export const getIngredientsFromSchema = () => {
  const ingredients = self.getPropertyFromSchema('recipeIngredient');
  if (!ingredients) return '';

  if (typeof ingredients === 'string') return ingredients;
  if (typeof ingredients[0] === 'string') return ingredients.join('\n');
  if (ingredients[0] && typeof ingredients[0].text === 'string') {
    return ingredients.map((ingredient) => ingredient.text).join('\n');
  }

  return '';
};
