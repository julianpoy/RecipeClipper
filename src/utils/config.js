export const generateConfig = (options) => {
  // Default structure
  const config = {
    window: null,
    options: {
      window: null,
      mlDisable: null,
      mlModelEndpoint: null,
      mlClassifyEndpoint: null,
    },
  };

  // Window-based options
  /* istanbul ignore next */
  try {
    if (window) {
      config.window = window;
      config.options.window = window;
      config.options.mlDisable = window.RC_ML_DISABLE || null;
      config.options.mlModelEndpoint = window.RC_ML_MODEL_ENDPOINT || null;
      config.options.mlClassifyEndpoint = window.RC_ML_CLASSIFY_ENDPOINT || null;
    }
  } catch (_) {
    // Do nothing
  }

  // Argument-based options
  if (options) {
    config.window = options.window || config.window;
    config.options = {
      ...config.options,
      ...options,
    };
  }

  return config;
};
