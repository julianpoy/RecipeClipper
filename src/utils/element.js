export const getClassNamesMatching = (classNamePartial) => {
  const classRegExp = new RegExp(`class="((\\w|\\s|-)*${classNamePartial}(\\w|\\s|-)*)"`, 'gi');
  const matches = document.body.innerHTML.matchAll(classRegExp);

  return Array.from(new Set(Array.from(matches, (match) => match[1])));
};

export const softMatchElementsByClass = (classNamePartial) => {
  const classNames = getClassNamesMatching(classNamePartial);

  return classNames
    .map((className) => Array.from(document.getElementsByClassName(className)))
    .flat();
};

export const applyLIBlockStyling = (element) => {
  Array.from(element.querySelectorAll('li')).forEach((li) => {
    li.style.display = 'block';
  });

  return element;
};

export const grabLongestMatchByClasses = (preferredClassNames, fuzzyClassNames) => {
  const exactMatches = preferredClassNames
    .map((className) => Array.from(document.getElementsByClassName(className)))
    .flat();
  const fuzzyMatches = fuzzyClassNames
    .map((className) => softMatchElementsByClass(className))
    .flat();

  return (exactMatches.length > 0 ? exactMatches : fuzzyMatches)
    .map((element) => applyLIBlockStyling(element))
    .map((element) => element.innerText.trim())
    .reduce((max, match) => (match.length > max.length ? match : max), '');
};

export const isImg = (element) => element.tagName.toLowerCase().trim() === 'img';
export const isPicture = (element) => element.tagName.toLowerCase().trim() === 'picture';

export const getImgElementsWithin = (element) => {
  const matchedImgElements = [];
  if (isImg(element)) matchedImgElements.push(element);
  matchedImgElements.push(...element.querySelectorAll('img'));
  return matchedImgElements;
};

export const getAttrIfExists = (el, attrName) => {
  if (el.attributes[attrName]) return el.attributes[attrName].value;
  return '';
};

export const getSrcFromImage = (img) => {
  if (!img) return '';

  const closestSrc = getAttrIfExists(img, 'data-src') || getAttrIfExists(img, 'data-lazy-src') || img.currentSrc || img.src;
  return closestSrc || '';
};

export const isValidImage = (element) => isImg(element)
    && !!getSrcFromImage(element)
    && element.complete // Filter images that haven't completely loaded
    && element.naturalWidth > 0 // Filter images that haven't loaded correctly
    && element.naturalHeight > 0;

export const getImageDimensions = (element) => {
  const parent = element.parentNode;
  const isParentPicture = parent && isPicture(parent);
  const offsetHeight = isParentPicture
    ? Math.max(element.offsetHeight, parent.offsetHeight) : element.offsetHeight;
  const offsetWidth = isParentPicture
    ? Math.max(element.offsetWidth, parent.offsetWidth) : element.offsetWidth;

  return {
    offsetHeight,
    offsetWidth,
  };
};

export const grabLargestImage = () => {
  const matches = document.querySelectorAll('img');

  return [...matches]
    .filter((element) => isValidImage(element))
    .reduce((max, element) => {
      const { offsetWidth, offsetHeight } = getImageDimensions(element);

      // Do not use images smaller than 200x200
      if (offsetWidth < 200 && offsetHeight < 200) return max;

      const elTotalPx = offsetHeight * offsetWidth;
      const maxTotalPx = max ? (max.offsetHeight * max.offsetWidth) : 0;

      return elTotalPx > maxTotalPx ? element : max;
    }, null);
};

export const closestToRegExp = (regExp) => {
  const match = document.body.innerText.match(regExp);
  if (!match) return '';
  return match[0];
};

export const grabRecipeTitleFromDocumentTitle = () => document.title.split(/ - |\|/)[0].trim();

export const grabSourceFromDocumentTitle = () => (document.title.split(/ - |\|/)[1] || '').trim();
