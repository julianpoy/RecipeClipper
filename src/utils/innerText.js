export const getInnerText = (element) => element.innerText || element.textContent;
