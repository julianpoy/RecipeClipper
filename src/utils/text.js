import { badWords } from '../constants/badWords';
import {
  matchStep, matchFieldTitles, matchScale, matchSpecialChracters,
} from '../constants/regex';

export const capitalizeEachWord = (textBlock) => textBlock.split(' ').map((word) => `${word.charAt(0).toUpperCase()}${word.substring(1)}`).join(' ');

export const removeSpecialCharacters = (line) => line.replace(matchSpecialChracters, '').trim();

// Undesired words
export const isBadWord = (line) => badWords.includes(removeSpecialCharacters(line).toLowerCase());

// Digits and steps that sit on their own lines
export const isStep = (line) => removeSpecialCharacters(line).match(matchStep);

// Scale buttons/interface that was picked up accidentally (4x)
export const isScale = (line) => removeSpecialCharacters(line).match(matchScale);

export const removeFieldTitles = (line) => line.replace(matchFieldTitles, '').trim();

// Format capitalized lines "HEADER:" as [Header] or "for the xyz" as [For The Xyz]
export const formatHeaders = (line) => (
  line.trim().match(/^([A-Z] *)+:?$/) || line.trim().match(/^for the ([a-z] *)+:?$/i)
    ? `[${capitalizeEachWord(line.trim().toLowerCase()).replace(':', '')}]` : line
);

export const cleanKnownWords = (textBlock) => textBlock.split('\n')
  .map((line) => line.trim())
  .filter((line) => removeSpecialCharacters(line).length)
  .filter((line) => !isBadWord(line))
  .filter((line) => !isStep(line))
  .filter((line) => !isScale(line))
  .map((line) => removeFieldTitles(line))
  .map((line) => formatHeaders(line))
  .join('\n');

export const format = {
  imageURL: (val) => val.trim(),
  title: (val) => capitalizeEachWord(val.trim().toLowerCase()),
  description: (val) => (val.length > 300 ? '' : cleanKnownWords(val)),
  source: (val) => val.trim(),
  yield: (val) => (val.length > 30 ? '' : capitalizeEachWord(cleanKnownWords(val).trim().toLowerCase())),
  activeTime: (val) => (val.length > 30 ? '' : capitalizeEachWord(cleanKnownWords(val).trim().toLowerCase())),
  totalTime: (val) => (val.length > 30 ? '' : capitalizeEachWord(cleanKnownWords(val).trim().toLowerCase())),
  ingredients: (val) => cleanKnownWords(val),
  instructions: (val) => cleanKnownWords(val),
  notes: (val) => cleanKnownWords(val),
};
