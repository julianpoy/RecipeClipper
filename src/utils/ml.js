// Self import for mock
import * as self from './ml';
import { getInnerText } from './innerText';
import {
  applyLIBlockStyling,
} from './element';
import {
  ingredientSectionHeader,
  instructionSectionHeader,
} from '../constants/regex';

export const getDocumentContent = (config) => {
  const { body } = config.window.document;
  applyLIBlockStyling(body);

  return getInnerText(body).split('\n').map((line) => line.trim()).filter((line) => line);
};

export const findPotentialSetsByHeader = (config, headerRegexp) => {
  const content = self.getDocumentContent(config);

  return content.filter((line) => line.match(headerRegexp))
    .map((line) => content.slice(content.indexOf(line) + 1));
};

export const loadModel = async (config) => {
  const modelUrl = config.options.mlModelEndpoint;
  if (!modelUrl) throw new Error('You must provide window.RC_ML_MODEL_ENDPOINT or options.mlModelEndpoint to use local classification');
  return config.window.tf.loadLayersModel(modelUrl);
};

export const mlClassifyLocal = async (config, lines) => {
  const model = await self.loadModel(config);
  const useModel = await config.window.use.load();

  const predictions = [];
  for (let i = 0; i < lines.length; i += 1) {
    // eslint-disable-next-line no-await-in-loop
    const encodedData = await useModel.embed([lines[i]]);
    const prediction = model.predict(encodedData).dataSync();
    predictions.push(prediction);
  }

  return predictions;
};

export const mlClassifyRemote = async (config, lines) => {
  const remote = config.options.mlClassifyEndpoint;
  if (!remote) throw new Error('You must provide window.RC_ML_CLASSIFY_ENDPOINT or options.mlClassifyEndpoint to use remote classification');

  const response = await config.window.fetch(remote, {
    method: 'POST',
    body: JSON.stringify({
      sentences: lines,
    }),
    headers: {
      'Content-Type': 'application/json',
    },
  });

  return response.json();
};

export const mlClassify = async (config, lines) => {
  const isTFJSAvailable = config.window.tf && config.window.tf.loadLayersModel;
  const isUSEAvailable = config.window.use && config.window.use.load;

  if (isTFJSAvailable && isUSEAvailable) return self.mlClassifyLocal(config, lines);
  return self.mlClassifyRemote(config, lines);
};

export const mlFilter = async (config, lines, type) => {
  const predictions = await self.mlClassify(config, lines);

  let lastType = -1;
  const filteredOutput = [];
  for (let i = 0; i < lines.length; i += 1) {
    const predictedType = predictions[i].indexOf(Math.max(...predictions[i])) + 1;

    // Allow one line of error to be included
    // (say, an ingredient header or a stray bit of formatting)
    // Also allow some non-ingredient text at the very start (say, a section header)
    if (predictedType === type || lastType === type || i === 0) {
      filteredOutput.push(lines[i]);
      lastType = predictedType;
    } else {
      break;
    }
  }

  // Remove last element if it is not of the desired type
  // (consider our allowed buffer above may insert one undesired item at end)
  if (filteredOutput.length > 0 && lastType !== type) {
    filteredOutput.pop();
  }

  return filteredOutput;
};

export const findFullSearch = async (config, type) => {
  const content = self.getDocumentContent(config);

  const predictions = await self.mlClassify(config, content);

  const { groups, workingGroup } = content.reduce((acc, line, idx) => {
    const predictedType = predictions[idx].indexOf(Math.max(...predictions[idx])) + 1;

    const lastType = acc.workingGroup
      && acc.workingGroup.length > 0 ? acc.workingGroup[acc.workingGroup.length - 1].type : -1;

    // Allow one line of error to be included in group
    // (say, an ingredient header or a stray bit of formatting)
    // Also allow some non-ingredient text at the very start (say, a section header)
    if (predictedType === type || lastType === type || idx === 0) {
      acc.workingGroup = acc.workingGroup || [];
      acc.workingGroup.push({
        text: line,
        type: predictedType,
      });
    } else {
      // Group is considered complete
      if (acc.workingGroup && acc.workingGroup.length > 0) acc.groups.push(acc.workingGroup);
      acc.workingGroup = null;
    }

    return acc;
  }, {
    workingGroup: null,
    groups: [],
  });

  // If leftover working group (if ingredients/instructions end at very end of page)
  if (workingGroup) groups.push(workingGroup);

  groups.forEach((group) => {
    // Remove last element if it is not of the desired type
    // (consider our allowed buffer above may insert one undesired item at end)
    if (group.length > 0 && group[group.length - 1].type !== type) {
      group.pop();
    }
  });

  return groups
    .map((group) => group.map((item) => item.text))
    .map((group) => group.join('\n'))
    .reduce((a, b) => (a.length > b.length ? a : b), '');
};

export const findByHeader = async (config, type) => {
  const headerRegexp = type === 1 ? ingredientSectionHeader : instructionSectionHeader;
  const potentialSets = self.findPotentialSetsByHeader(config, headerRegexp);

  const sets = [];
  for (let i = 0; i < potentialSets.length; i += 1) {
    const potentialSet = potentialSets[i];
    // eslint-disable-next-line no-await-in-loop
    const set = await self.mlFilter(config, potentialSet, type);
    sets.push(set);
  }

  return sets.map((set) => set.join('\n'))
    .reduce((a, b) => (a.length > b.length ? a : b), '');
};

export const find = async (config, type) => {
  if (config.options.mlDisable) return '';

  const result = await self.findByHeader(config, type) || await self.findFullSearch(config, type);

  return result;
};

// Type 1 for ingredients
// Type 2 for instructions
// Others to be implemented in future...
export const grabByMl = async (config, type) => {
  try {
    return await self.find(config, type);
  } catch (e) {
    if (config.options.ignoreMLClassifyErrors) {
      return '';
    }

    throw e;
  }
};
