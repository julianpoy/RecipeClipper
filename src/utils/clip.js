import { format } from './text';
import {
  getSrcFromImage,
  grabLargestImage,
  grabLongestMatchByClasses,
  closestToRegExp,
  grabRecipeTitleFromDocumentTitle,
  grabSourceFromDocumentTitle,
} from './element';
import ClassMatchers from '../constants/classMatchers';
import {
  matchYield,
  matchActiveTime,
  matchTotalTime,
} from '../constants/regex';
import {
  grabByMl
} from './ml';

export const clipImageURL = () => format.imageURL(
  getSrcFromImage(grabLargestImage()),
);

export const clipTitle = () => format.title(
  grabLongestMatchByClasses(...ClassMatchers.title) || grabRecipeTitleFromDocumentTitle(),
);

export const clipDescription = () => format.description(
  grabLongestMatchByClasses(...ClassMatchers.description),
);

export const clipSource = () => format.source(
  grabSourceFromDocumentTitle() || window.location.hostname,
);

export const clipYield = () => format.yield(
  grabLongestMatchByClasses(...ClassMatchers.yield)
  || closestToRegExp(matchYield).replace('\n', ''),
);

export const clipActiveTime = () => format.activeTime(
  grabLongestMatchByClasses(...ClassMatchers.activeTime)
  || closestToRegExp(matchActiveTime).replace('\n', ''),
);

export const clipTotalTime = () => format.totalTime(
  grabLongestMatchByClasses(...ClassMatchers.totalTime)
  || closestToRegExp(matchTotalTime).replace('\n', ''),
);

export const clipIngredients = async () => format.ingredients(
  grabLongestMatchByClasses(...ClassMatchers.ingredients)
  || await grabByMl(1),
);

export const clipInstructions = async () => format.instructions(
  grabLongestMatchByClasses(...ClassMatchers.instructions)
  || await grabByMl(2),
);

export const clipNotes = () => format.notes(
  grabLongestMatchByClasses(...ClassMatchers.notes),
);
