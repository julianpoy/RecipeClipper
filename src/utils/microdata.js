import * as self from './microdata';
import { getInnerText } from './innerText';
import { getLongestString } from './text';

export const getLongestSingleElementTextForQueries = (window, queries) => {
  const vals = queries
    .map((query) => [...window.document.querySelectorAll(query)])
    .flat()
    .map((el) => getInnerText(el));

  return getLongestString(vals);
};

export const getLongestTextGroupForQueries = (window, queries) => {
  const vals = queries
    .map((query) =>
         [...window.document.querySelectorAll(query)]
           .map((el) => getInnerText(el))
           .join('\n');
    );

  return getLongestString(vals);
};

export const getActiveTimeFromMicrodata = (window) => self.getLongestSingleElementTextForQueries(window, [
  '[itemProp=prepTime]',
]);

export const getTotalTimeFromMicrodata = (window) => self.getLongestSingleElementTextForQueries(window, [
  '[itemProp=totalTime]',
]);

export const getYieldFromMicrodata = (window) => self.getLongestSingleElementTextForQueries(window, [
  '[itemProp=recipeYield]',
]);

export const getInstructionsFromMicrodata = (window) => self.getLongestTextGroupForQueries(window, [
  '[itemProp=recipeInstructions]',
  '[itemProp=recipeInstruction]',
  '[itemProp=instructions]',
  '[itemProp=instruction]',
]);

export const getIngredientsFromMicrodata = (window) => self.getLongestTextGroupForQueries(window, [
  '[itemProp=recipeIngredients]',
  '[itemProp=recipeIngredient]',
  '[itemProp=ingredients]',
  '[itemProp=ingredient]',
]);
