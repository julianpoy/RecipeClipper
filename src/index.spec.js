import { clipRecipe } from './index';
import * as ClipUtils from './utils/clip';

describe('clip', () => {
  let clipImageURLSpy, clipTitleSpy, clipDescriptionSpy, clipSourceSpy, clipYieldSpy;
  let clipTotalTimeSpy, clipIngredientsSpy, clipInstructionsSpy, clipNotesSpy, clipActiveTimeSpy;
  let outputVal;

  beforeAll(() => {
    clipImageURLSpy = jest.spyOn(ClipUtils, 'clipImageURL').mockReturnValue('imageURL');
    clipTitleSpy = jest.spyOn(ClipUtils, 'clipTitle').mockReturnValue('title');
    clipDescriptionSpy = jest.spyOn(ClipUtils, 'clipDescription').mockReturnValue('description');
    clipSourceSpy = jest.spyOn(ClipUtils, 'clipSource').mockReturnValue('source');
    clipYieldSpy = jest.spyOn(ClipUtils, 'clipYield').mockReturnValue('yield');
    clipActiveTimeSpy = jest.spyOn(ClipUtils, 'clipActiveTime').mockReturnValue('activeTime');
    clipTotalTimeSpy = jest.spyOn(ClipUtils, 'clipTotalTime').mockReturnValue('totalTime');
    clipIngredientsSpy = jest.spyOn(ClipUtils, 'clipIngredients').mockReturnValue('ingredients');
    clipInstructionsSpy = jest.spyOn(ClipUtils, 'clipInstructions').mockReturnValue('instructions');
    clipNotesSpy = jest.spyOn(ClipUtils, 'clipNotes').mockReturnValue('notes');

    outputVal = clipRecipe();
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('calls clippers for subfields', () => {
    expect(clipImageURLSpy).toBeCalled();
    expect(clipTitleSpy).toBeCalled();
    expect(clipDescriptionSpy).toBeCalled();
    expect(clipSourceSpy).toBeCalled();
    expect(clipYieldSpy).toBeCalled();
    expect(clipActiveTimeSpy).toBeCalled();
    expect(clipTotalTimeSpy).toBeCalled();
    expect(clipIngredientsSpy).toBeCalled();
    expect(clipInstructionsSpy).toBeCalled();
    expect(clipNotesSpy).toBeCalled();
  });

  it('returns result matching snapshot', () => {
    expect(outputVal).toMatchSnapshot();
  });
});
