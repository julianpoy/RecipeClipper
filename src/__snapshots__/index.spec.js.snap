// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clip returns result matching snapshot 1`] = `
Object {
  "activeTime": "activeTime",
  "description": "description",
  "imageURL": "imageURL",
  "ingredients": "ingredients",
  "instructions": "instructions",
  "notes": "notes",
  "source": "source",
  "title": "title",
  "totalTime": "totalTime",
  "yield": "yield",
}
`;
